lightning-transaction-sync/src/tests.rs

```diff
@@ -0,0 +1,242 @@
+use crate::esplora::EsploraSyncClient;
+use lightning::chain::{Confirm, Filter};
+use lightning::chain::transaction::TransactionData;
+use lightning::util::logger::{Logger, Record};
+
+use electrsd::{bitcoind, bitcoind::BitcoinD, ElectrsD};
+use bitcoin::{Amount, Txid, BlockHash, BlockHeader};
+use bitcoin::blockdata::constants::genesis_block;
+use bitcoin::network::constants::Network;
+use electrsd::bitcoind::bitcoincore_rpc::bitcoincore_rpc_json::AddressType;
+use bitcoind::bitcoincore_rpc::RpcApi;
+use electrum_client::ElectrumApi;
+
+use once_cell::sync::OnceCell;
+
+use std::env;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::collections::{HashMap, HashSet};
+
+static BITCOIND: OnceCell<BitcoinD> = OnceCell::new();
+static ELECTRSD: OnceCell<ElectrsD> = OnceCell::new();
+static PREMINE: OnceCell<()> = OnceCell::new();
+static MINER_LOCK: Mutex<()> = Mutex::new(());
+
+fn get_bitcoind() -> &'static BitcoinD {
+	BITCOIND.get_or_init(|| {
+		let bitcoind_exe =
+			env::var("BITCOIND_EXE").ok().or_else(|| bitcoind::downloaded_exe_path().ok()).expect(
+				"you need to provide an env var BITCOIND_EXE or specify a bitcoind version feature",
+				);
+		let mut conf = bitcoind::Conf::default();
+		conf.network = "regtest";
+		BitcoinD::with_conf(bitcoind_exe, &conf).unwrap()
+	})
+}
+
+fn get_electrsd() -> &'static ElectrsD {
+	ELECTRSD.get_or_init(|| {
+		let bitcoind = get_bitcoind();
+		let electrs_exe =
+			env::var("ELECTRS_EXE").ok().or_else(electrsd::downloaded_exe_path).expect(
+				"you need to provide env var ELECTRS_EXE or specify an electrsd version feature",
+			);
+		let mut conf = electrsd::Conf::default();
+		conf.http_enabled = true;
+		conf.network = "regtest";
+		ElectrsD::with_conf(electrs_exe, &bitcoind, &conf).unwrap()
+	})
+}
+
+fn generate_blocks_and_wait(num: usize) {
+	let cur_height = get_bitcoind().client.get_block_count().unwrap();
+	generate_blocks(num);
+	wait_for_block(cur_height as usize + num);
+}
+
+fn generate_blocks(num: usize) {
+	let address = get_bitcoind().client.get_new_address(Some("test"), Some(AddressType::Legacy)).unwrap();
+	let _block_hashes = get_bitcoind().client.generate_to_address(num as u64, &address).unwrap();
+}
+
+fn wait_for_block(min_height: usize) {
+	let mut header = get_electrsd().client.block_headers_subscribe().unwrap();
+	loop {
+		if header.height >= min_height {
+			break;
+		}
+		header = exponential_backoff_poll(|| {
+			get_electrsd().trigger().unwrap();
+			get_electrsd().client.ping().unwrap();
+			get_electrsd().client.block_headers_pop().unwrap()
+		});
+	}
+}
+
+fn exponential_backoff_poll<T, F>(mut poll: F) -> T
+where
+	F: FnMut() -> Option<T>,
+{
+	let mut delay = Duration::from_millis(64);
+	loop {
+		match poll() {
+			Some(data) => break data,
+			None if delay.as_millis() < 512 => delay = delay.mul_f32(2.0),
+			None => {}
```

Comment:
Do we ever need to give up? Or will we just fail if the closure unwraps an `Err`?